File: apps/api/routers/taxonomy.py
"""Taxonomy router (ADR-028 increment 2a) — serve the field forest, edit the curated hierarchy.

Read model + the write endpoints over the ``knowledge/taxonomy.py`` seam. Edits the *taxonomy*
(hierarchy edges + document→field links), never the concept vocabulary itself (ADR-017 A1 / ADR-019
D11: a dedicated taxonomy surface owns tree edits; concept create/rename/delete stay elsewhere).
$0, zero-LLM — the whole increment is deterministic sidecar reads/writes.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException

from apps.api.models import (
    FieldDetailPayload,
    HierarchyEdgeRequest,
    TaxonomyViewPayload,
)

router = APIRouter()


@router.get("/api/taxonomy")
def get_taxonomy() -> TaxonomyViewPayload:
    """The curated field forest + per-field coverage + corpus totals.

    Not 404 on an empty forest: the ANZSRC trunk is bundled seed data, so "no fields yet" is a
    legitimate unseeded state (run ``scripts/seed_taxonomy --apply``), not a missing artifact.
    Every rollup is 0 until concepts/documents are attached — the honest zero-state."""
    from doc_assistant.knowledge.taxonomy_view import load_taxonomy_view

    return TaxonomyViewPayload.from_view(load_taxonomy_view())


@router.get("/api/taxonomy/fields/{field_id}")
def get_field_detail(field_id: str) -> FieldDetailPayload:
    """One field's directly-attached concepts + documents + rollup counts.

    404 when ``field_id`` is not a ``kind="domain"`` node — a wrong id or a concept id — as
    distinct from a real-but-empty field, which returns a detail with empty member lists."""
    from doc_assistant.knowledge.taxonomy_view import load_field_detail

    detail = load_field_detail(field_id)
    if detail is None:
        raise HTTPException(status_code=404, detail=f"no taxonomy field with id {field_id!r}")
    return FieldDetailPayload.from_detail(detail)


@router.post("/api/taxonomy/hierarchy", status_code=201)
def add_hierarchy(body: HierarchyEdgeRequest) -> dict[str, Any]:
    """Add one curated hierarchy edge (``source --type--> target``). Idempotent on the unique key.

    Attaching a concept to a field is just an ``in_field`` edge from the concept to the domain
    node — same endpoint. ``409`` if the edge would create a cycle (ADR-028 D3); ``404`` if an id
    is not a concept/domain row; ``400`` on a bad ``type`` (guarded by the request model too)."""
    from sqlalchemy import select

    from doc_assistant.db.models import Concept
    from doc_assistant.db.session import session_scope
    from doc_assistant.knowledge.taxonomy import TaxonomyCycleError, add_hierarchy_edge

    with session_scope() as session:
        present = set(
            session.execute(
                select(Concept.id).where(Concept.id.in_([body.source_id, body.target_id]))
            )
            .scalars()
            .all()
        )
        for cid in (body.source_id, body.target_id):
            if cid not in present:
                raise HTTPException(status_code=404, detail=f"no concept/field with id {cid!r}")
        try:
            add_hierarchy_edge(session, body.source_id, body.target_id, body.type)
        except TaxonomyCycleError as e:
            raise HTTPException(status_code=409, detail=str(e)) from e
        except ValueError as e:  # unknown edge type (the model narrows this, defence in depth)
            raise HTTPException(status_code=400, detail=str(e)) from e
    return {
        "ok": True,
        "source_id": body.source_id,
        "target_id": body.target_id,
        "type": body.type,
    }


@router.delete("/api/taxonomy/hierarchy")
def remove_hierarchy(body: HierarchyEdgeRequest) -> dict[str, int]:
    """Remove a curated hierarchy edge by its unique key. Idempotent (``removed`` 0 if absent)."""
    from doc_assistant.db.session import session_scope
    from doc_assistant.knowledge.taxonomy import remove_hierarchy_edge

    with session_scope() as session:
        removed = remove_hierarchy_edge(session, body.source_id, body.target_id, body.type)
    return {"removed": removed}


@router.post("/api/taxonomy/documents/{document_id}/fields/{field_id}", status_code=201)
def attach_document(document_id: str, field_id: str) -> dict[str, Any]:
    """Attach a document to a taxonomy field (a ``kind="domain"`` node). Idempotent per pair.

    ``404`` if the document does not exist; ``400`` if the target is not a domain field
    (``NotADomainError``)."""
    from sqlalchemy import select

    from doc_assistant.db.models import Document
    from doc_assistant.db.session import session_scope
    from doc_assistant.knowledge.taxonomy import NotADomainError, attach_document_field

    with session_scope() as session:
        exists = session.execute(
            select(Document.id).where(Document.id == document_id)
        ).scalar_one_or_none()
        if exists is None:
            raise HTTPException(status_code=404, detail=f"no document with id {document_id!r}")
        try:
            attach_document_field(session, document_id, field_id)
        except NotADomainError as e:
            raise HTTPException(status_code=400, detail=str(e)) from e
    return {"ok": True, "document_id": document_id, "field_id": field_id}

File: apps/desktop/src/lib/core/api/taxonomy.ts
// Thin `taxonomy` client for the desktop API — fetch + parsing only, no business logic.
// Pairs with apps/api/routers/taxonomy.py and apps/api/models/taxonomy.py; see
// docs/architecture.md, section "apps/ — the domain spine".

import { API_BASE, errorDetail } from './_base'
import type {
  FieldDetail,
  HierarchyEdgeRequest,
  TaxonomyView,
} from '../types'

/** The field forest + per-field coverage + corpus totals. 200 with zeros on an unseeded/empty
 *  forest (the ANZSRC trunk is bundled data — an empty forest is a legit state, not a 404). */
export async function getTaxonomy(): Promise<TaxonomyView> {
  const r = await fetch(`${API_BASE}/api/taxonomy`)
  if (!r.ok) throw new Error(await errorDetail(r, 'taxonomy'))
  return (await r.json()) as TaxonomyView
}
/** One field's directly-attached concepts + documents + rollup counts. 404 when `fieldId` is not a
 *  domain node (a wrong id or a concept id) — as distinct from a real-but-empty field. */
export async function getFieldDetail(fieldId: string): Promise<FieldDetail> {
  const r = await fetch(`${API_BASE}/api/taxonomy/fields/${encodeURIComponent(fieldId)}`)
  if (!r.ok) throw new Error(await errorDetail(r, 'field detail'))
  return (await r.json()) as FieldDetail
}
/** Add one curated hierarchy edge (`source --type--> target`). Attaching a concept to a field is
 *  just an `in_field` edge from the concept to the domain node — same endpoint. Surfaces the
 *  backend's 409 (cycle) / 404 (missing id) / 400 (bad type) `detail` message. */
export async function addHierarchyEdge(body: HierarchyEdgeRequest): Promise<void> {
  const r = await fetch(`${API_BASE}/api/taxonomy/hierarchy`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!r.ok) throw new Error(await errorDetail(r, 'add hierarchy edge'))
}
/** Remove a curated hierarchy edge by its unique key. Idempotent (`removed` 0 if absent). The
 *  client's first DELETE with a JSON body — `fetch` allows it; set the JSON content-type. */
export async function removeHierarchyEdge(body: HierarchyEdgeRequest): Promise<{ removed: number }> {
  const r = await fetch(`${API_BASE}/api/taxonomy/hierarchy`, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!r.ok) throw new Error(await errorDetail(r, 'remove hierarchy edge'))
  return (await r.json()) as { removed: number }
}
/** Attach a document to a taxonomy field (a domain node). Idempotent per pair. 404 for an unknown
 *  document; 400 when the target is not a domain field (`NotADomainError`). */
export async function attachDocumentField(docId: string, fieldId: string): Promise<void> {
  const r = await fetch(
    `${API_BASE}/api/taxonomy/documents/${encodeURIComponent(docId)}/fields/${encodeURIComponent(fieldId)}`,
    { method: 'POST' },
  )
  if (!r.ok) throw new Error(await errorDetail(r, 'attach document to field'))
}
/** Remove a document→field link. Idempotent (`removed` 0 if absent). The reject half of the
 *  auto-propose loop (ADR-028 D8) — and equally undoes a curated attach. */
export async function detachDocumentField(
  docId: string,
  fieldId: string,
): Promise<{ removed: number }> {
  const r = await fetch(
    `${API_BASE}/api/taxonomy/documents/${encodeURIComponent(docId)}/fields/${encodeURIComponent(fieldId)}`,
    { method: 'DELETE' },
  )
  if (!r.ok) throw new Error(await errorDetail(r, 'detach document from field'))
  return (await r.json()) as { removed: number }
}

/** Pull a human message out of a FastAPI error body, falling back to the status code. `detail` is
 *  usually a string, but the selective-ingest 400 sends `{error, offenders}` — render that too. */
